DSpace/DSpace
# Duplicate Detection ported from DSpace CRIS 7

This pull request ports the Duplicate Detection framework from DSpace CRIS 7 into DSpace.

*Attribution: Ported from DSpace-CRIS and extended by The Library Code with support of Technische Universität Berlin*

Angular PR: https://github.com/DSpace/dspace-angular/pull/1732
REST Contract PR: [Link](https://github.com/DSpace/RestContract/pull/198)

**To test**
- Make sure both the dspace-angular and DSpace PRs are applied
- Enable the ‘dedup’ event consumer in dspace.cfg and add the ‘detect-duplicate’ step to a submission process in item-submission.xml
- For best results, configure autosave on the dc.title field and any others you’re planning to test (eg. dc.title, dc.identifier.doi, dc.identifier.isbn or so on). This means that whenever a change is saved to dc.title, the duplicate detection section will update with a new list of potential matches
- Submit an item to a collection with the detect-duplicate step enabled. Enter some text in the title so that it matches an existing item in the archive (current user must have READ rights) and tab out of the field to force a save (see autosave recommendation -- if not using, complete a section eg file upload or licence grant). A potential duplicate should appear at the detect-duplicate section.
- Change the title to something completely different and tab out again. The duplicate should disappear.
- Change the text so that it is off by *one* character (more specifically, off by 1 levenshtein distance) and tab out again. The duplicate should reappear. Eg. Sample vs S_mple.
- Change the text so that it is still similar but off by more than 1 eg. Sample vs Sam__e. The duplicate should disappear again (note: max distance is configurable but default is 1)
- With one more more duplicates showing, register a decision using the buttons provided. Note that these don't do much currently - they allow a decision to be recorded where the submitter acknowledges a duplicate but still wishes to submit, and also allows the submitter or reviewer to get rid of false positives.

Tests can be repeated as a workflow reviewer/editor. Decisions made by submitters are able to be overridden / rejected duplicates can be seen by reviewers. This is configured true by default.

I appreciate this is a large change with a lot of parts, so please contact me in this Github thread or on Slack for any testing assistance needed.

**Notes for PR reviewers**
- There is a new ‘dedup’ table in the database, this should be added by flyway
- There is a new ‘dedup’ Solr core
- The ‘dedup’ consumer is not in the default event consumers by default, so it’ll need to be added for practical testing
- The fuzzy search signature is commented out of the deduplication.xml spring config by default, so uncomment that to test fuzzy matches (the default MD5 signatures for exact metadata matches still work)

**Notes for 4Science devs (regarding changes from CRIS)**
- SearchDeduplication was renamed to DeduplicationPluginService as that seemed more fitting to what it actually does
- tmpMapFilter, tmpFilter, other similar names of maps and lists were renamed to suit their purpose (eg. signatureSearchFilters)
- A lot of unused methods, variables, classes were stripped out during the port to keep code lean and easier to review. If some were partial implementations of future work, we can bring them in with a subsequent PR

As I’ve added an authorization check (configurable, enabled by default) so that the user needs read access to each item in the potential matches list, I manually disabled it for some integration tests and leave it for others, and explicitly test it in a new integration test

I’ve added javadoc and most is pretty straightforward but please correct me if I misunderstood terms such as ‘fake', ‘reader’

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] If my PR modifies the REST API, I've linked to the REST Contract page [here](https://github.com/DSpace/RestContract/pull/198)
